webapp/src/ts/services/contact-save.service.ts

```diff
@@ -0,0 +1,191 @@
+    return childData.map(child => {
+      child.parent = this.extractLineageService.extract(doc);
+      return this.prepare(child);
+    });
+  }
+
+  private extractIfRequired(name, value) {
+    return this.CONTACT_FIELD_NAMES.includes(name) ? this.extractLineageService.extract(value) : value;
+  }
+
+  // Mutates the passed doc to attach prepared siblings, and returns all
+  // prepared siblings to be persisted.
+  // This will (on a correctly configured form) attach the full parent to
+  // doc, and in turn siblings. See internal comments.
+  private prepareAndAttachSiblingDocs(doc, original, siblings) {
+    if (!doc._id) {
+      throw new Error('doc passed must already be prepared with an _id');
+    }
+
+    const preparedSiblings = [];
+    let promiseChain = Promise.resolve();
+
+    this.CONTACT_FIELD_NAMES.forEach(fieldName => {
+      let value = doc[fieldName];
+      if (_isObject(value)) {
+        value = doc[fieldName]._id;
+      }
+      if (!value) {
+        return;
+      }
+      if (value === 'NEW') {
+        const preparedSibling = this.prepare(siblings[fieldName]);
+
+        // by default all siblings are "person" types but can be overridden
+        // by specifying the type and contact_type in the form
+        if (!preparedSibling.type) {
+          preparedSibling.type = 'person';
+        }
+
+        if (preparedSibling.parent === 'PARENT') {
+          delete preparedSibling.parent;
+          // Cloning to avoid the circular references
+          doc[fieldName] = { ...preparedSibling };
+          // Because we're assigning the actual doc reference, the dbService.get.get
+          // to attach the full parent to the doc will also attach it here.
+          preparedSibling.parent = doc;
+        } else {
+          doc[fieldName] = this.extractIfRequired(fieldName, preparedSibling);
+        }
+
+        preparedSiblings.push(preparedSibling);
+      } else if (original?.[fieldName]?._id === value) {
+        doc[fieldName] = original[fieldName];
+      } else {
+        promiseChain = promiseChain.then(() => {
+          return this.dbService
+            .get()
+            .get(value)
+            .then((dbFieldValue) => {
+              // In a correctly configured form one of these will be the
+              // parent. This must happen before we attempt to run
+              // ExtractLineage on any siblings or repeats, otherwise they
+              // will extract an incomplete lineage
+              doc[fieldName] = this.extractIfRequired(fieldName, dbFieldValue);
+            });
+        });
+      }
+    });
+
+    return promiseChain.then(() => preparedSiblings);
+  }
+
+  save(form, docId, type) {
+    return Promise
+      .resolve()
+      .then(() => docId ? this.dbService.get().get(docId) : null)
+      .then(original => {
+        const submitted = this.enketoTranslationService.contactRecordToJs(form.getDataStr({ irrelevant: false }));
+        if (original) {
```

Comment:
This is like pre-preparing docs, is it possible to put this logic inside the `prepareSubmittedDocsForSave ` ? :)